v2/yarpcgrpc/outbound.go

```diff
@@ -46,83 +42,55 @@ import (
 
-// Outbound is a transport.UnaryOutbound.
+// Outbound sends YARPC requests over gRPC. It is recommended that services use
+// a single HTTP dialer to construct all HTTP outbounds, ensuring efficient
+// sharing of resources across the different outbounds.
 type Outbound struct {
-	once        *lifecycle.Once
-	lock        sync.Mutex
-	t           *Transport
-	peerChooser peer.Chooser
-	options     *outboundOptions
-}
+	// Chooser is a peer chooser for outbound requests.
+	Chooser yarpc.Chooser
 
-func newSingleOutbound(t *Transport, address string, options ...OutboundOption) *Outbound {
-	return newOutbound(t, peerchooser.NewSingle(hostport.PeerIdentifier(address), t), options...)
-}
+	// Dialer is an alternative to specifying a Chooser.
+	// The outbound will dial the address specified in the URL.
+	Dialer yarpc.Dialer
 
-func newOutbound(t *Transport, peerChooser peer.Chooser, options ...OutboundOption) *Outbound {
-	return &Outbound{
-		once:        lifecycle.NewOnce(),
-		t:           t,
-		peerChooser: peerChooser,
-		options:     newOutboundOptions(options),
-	}
-}
+	// URL specifies the template for the URL this outbound makes requests to.
+	// For yarpc.Chooser-based outbounds, the peer (host:port) spection of the
+	// URL may vary from call to call but the REST will remain unchanged.
+	// For single-peer outbounds, the URL will be used as-is.
+	URL *url.URL
 
-// Start implements transport.Lifecycle#Start.
-func (o *Outbound) Start() error {
-	return o.once.Start(o.peerChooser.Start)
+	// Tracer attaches a tracer for the outbound.
+	Tracer opentracing.Tracer
 }
 
-// Stop implements transport.Lifecycle#Stop.
-func (o *Outbound) Stop() error {
-	return o.once.Stop(o.peerChooser.Stop)
-}
-
-// IsRunning implements transport.Lifecycle#IsRunning.
-func (o *Outbound) IsRunning() bool {
-	return o.once.IsRunning()
-}
-
-// Transports implements transport.Inbound#Transports.
-func (o *Outbound) Transports() []transport.Transport {
-	return []transport.Transport{o.t}
-}
-
-// Chooser returns the peer.Chooser associated with this Outbound.
-func (o *Outbound) Chooser() peer.Chooser {
-	return o.peerChooser
-}
-
-// Call implements transport.UnaryOutbound#Call.
-func (o *Outbound) Call(ctx context.Context, request *transport.Request) (*transport.Response, error) {
-	if request == nil {
-		return nil, yarpcerrors.InvalidArgumentErrorf("request for grpc outbound was nil")
-	}
-	if err := o.once.WaitUntilRunning(ctx); err != nil {
-		return nil, intyarpcerrors.AnnotateWithInfo(yarpcerrors.FromError(err), "error waiting for grpc outbound to start for service: %s", request.Service)
+// Call implements yarpc.UnaryOutbound#Call.
+func (o *Outbound) Call(ctx context.Context, req *yarpc.Request, reqBuf *yarpc.Buffer) (*yarpc.Response, *yarpc.Buffer, error) {
+	if req == nil {
+		return nil, nil, yarpcerror.InvalidArgumentErrorf("request for grpc outbound was nil")
 	}
+
 	start := time.Now()
 
 	var responseBody []byte
 	var responseMD metadata.MD
```

Comment:
optionally, could use `var` grouping here:
```
var (
  responseBody []byte
  responseMD   metadata.MD
)
```

Reply:
Also, given that these variables are passed into `invoke` and mutated there, we could update this so that we initialize these variables in `invoke` and explicitly return the values. That way, we can read this as:
```
responseBody, responseMD, invokeErr := o.invoke(ctx, req, reqBuf, start)
```

Additionally, do we need to pass in the `time.Time` parameter? Can we not call `time.Now()` from within the `invoke` call, too?

That would simplify the `invoke` function signature further to:
```
responseBody, responseMD, invokeErr := o.invoke(ctx, req, reqBuf)
```

Reply:
good call, this is left over from the previous implementation